### ISSUE TITLE
`NavigationCamera` initial transition flies across the map

### ISSUE THREAD

**Issue by kmadsen** [mapbox/mapbox-navigation-android#4679]
**Mapbox Navigation SDK version:** com.mapbox.navigation:android:2.0.0-beta.20

### Issue description

As shown below, without the `NavigationCamera` the map loads immediately at the current location. This tells us that the `NavigationCamera` has a way to load at the current location without flying across the map.

There is also side by side comparisons of an integration with the `NavigationCamera`. One with no transition, and the other with a default transition. The video shows the initial transition starts with an ocean view. Removing the transition should not be needed to work around this issue.

### Actual behavior

Without the `NavigationCamera`

![output](https://user-images.githubusercontent.com/3021882/127701882-1d139e2d-8caa-43c4-acc9-714976ae19d9.gif)

With the `NavigationCamera`

| default transition | 0 duration transition | 
| -- | -- |
| ![output](https://user-images.githubusercontent.com/3021882/127701764-f1fb5fed-81aa-4585-9c9d-277f67b75cc6.gif) | ![output](https://user-images.githubusercontent.com/3021882/127701634-5ae5c594-54e2-4ea7-b337-12244713f87a.gif) |

### Expected behavior

The camera should initialize without flying across the map

### Code for creating 0 duration transition

``` kotlin
when (cameraMode) {
    CameraMode.IDLE -> navigationCamera.requestNavigationCameraToIdle()
    CameraMode.FOLLOWING -> navigationCamera.requestNavigationCameraToFollowing(
        stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
            .maxDuration(0)
            .build()
    )
    CameraMode.OVERVIEW -> navigationCamera.requestNavigationCameraToOverview(
        stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
            .maxDuration(0)
            .build()
    )
}
```

**Comment by kmadsen** [comment]
cc: @LukasPaczos @d-prukop

**Comment by samcrawford** [comment]
FWIW, I experienced the same behaviour on 1.6.1+. I found it was resolved by enabling Google's Fused location provider:

    implementation("com.google.android.gms:play-services-location:18.0.0")

Hope this helps in tracking down the bug.

**Comment by zugaldia** [comment]
@kmadsen is this issue Android Auto-specific?

**Comment by kmadsen** [comment]
> @ kmadsen is this issue Android Auto-specific?

@zugaldia no this is an issue when you build on the `NavigationCamera`. There are more videos here when investigating the issue https://github.com/mapbox/mapbox-navigation-android/pull/4687. It's hard to specify the initial transition.

The navigation-sdk camera example also has the issue of starting in the ocean. Here is the video of the camera example. The slower your GPS is, the longer you will be in the ocean.

https://user-images.githubusercontent.com/3021882/128365957-edc47286-64a6-463d-9d36-d1703d1cc6d2.mp4

**Comment by Guardiola31337** [comment]
For reference, noting that this is not new, we had the same issue in the legacy https://github.com/mapbox/mapbox-navigation-android/issues/1330 https://github.com/mapbox/mapbox-navigation-android/issues/1245 and fixed it in https://github.com/mapbox/mapbox-navigation-android/pull/1483

**Comment by kmadsen** [comment]
The fix from `@ danesfeder` looks good too. Giving an option for an initial position https://github.com/mapbox/mapbox-navigation-android/pull/1483. That created the best experience I could find so far over here https://github.com/mapbox/mapbox-navigation-android/pull/4687#issuecomment-893042836

But the API is also confusing there `.easeTo(..)` or `.flyTo(..)` needs direct access to the camera first. The higher level interface should include a way to handle this though.

**Comment by kmadsen** [comment]
Setting up the camera directly before setting up the NavigationCamera can give a nice experience though

https://user-images.githubusercontent.com/3021882/128379465-f718e9b0-cac5-40cc-baab-bff6b4c05209.mov

**Comment by LukasPaczos** [comment]
`NavigationCameraTransitionOptions` were exposed exactly to help resolve this problem.

The first-ever transition is recommended to be executed with 0 duration, exactly like you did in your example @kmadsen:
```
    navigationCamera.requestNavigationCameraToFollowing(
        stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
            .maxDuration(0)
            .build()
    )
```

Do you think we should do more than that?

**Comment by kmadsen** [comment]
> Do you think we should do more than that?

Yeah I think so. It took awhile to figure this all out. Also the Camera Example should have the recommended experience here.

I'm not able to resolve the problem using `NavigationCameraTransitionOptions`. The best solution i've found is to not update the `NavigationCamera` on the first location and only update the camera directly with a `MapPluginProviderDelegate`

``` kotlin
            if (!jumpedToFirstLocation) {
                jumpedToFirstLocation = true
                carMapSurface?.mapSurface?.camera?.flyTo(
                    CameraOptions.Builder()
                        .center(
                            Point.fromLngLat(
                                enhancedLocation.longitude,
                                enhancedLocation.latitude
                            )
                        )
                        .build(),
                    MapAnimationOptions.Builder().duration(0L).build()
                )
            } else {
                viewportDataSource.onLocationChanged(enhancedLocation)
                viewportDataSource.evaluate()
                navigationCamera.requestNavigationCameraToFollowing()
            }
```

I'd expect the `requestNavigationCameraToFollowing` to handle this though

**Comment by LukasPaczos** [comment]
> I'm not able to resolve the problem using NavigationCameraTransitionOptions.

When are you executing the transition? From the video, I suspect you're doing it after the style is loaded. You'd have to execute it before the style loads for the map to initialize at the desired location instead of over the ocean.

Here's an example that I'm running:

https://user-images.githubusercontent.com/16925074/128387690-e731a021-6d8c-4618-ab74-ed424d003db4.mp4

and the only block of code that I have is:
```kotlin
    private val locationObserver = object : LocationObserver {
        var firstLocationUpdateReceived = false

        override fun onRawLocationChanged(rawLocation: Location) {
            // not handled
        }

        override fun onEnhancedLocationChanged(
            enhancedLocation: Location,
            keyPoints: List<Location>
        ) {
            // update location puck's position on the map
            navigationLocationProvider.changePosition(
                location = enhancedLocation,
                keyPoints = keyPoints
            )

            // update camera position to account for new location
            viewportDataSource.onLocationChanged(enhancedLocation)
            viewportDataSource.evaluate()

            // if this is the first location update the activity has received,
            // it's best to immediately move the camera to the current user location
            if (!firstLocationUpdateReceived) {
                firstLocationUpdateReceived = true
                navigationCamera.requestNavigationCameraToOverview(
                    stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
                        .maxDuration(0) // instant transition
                        .build()
                )
            }
        }
    }
```

This _does not_ wait for the style to get loaded to call `requestNavigationCameraToOverview` with `0` duration.

Could you double-check this in your implementation so that we can rule out a bug? I agree that there's still room for improvement and possibly `NavigationCamera` could handle that automatically, not yet clear how we could achieve that though.

**Comment by LukasPaczos** [comment]
Of course, if you need to wait for a location update longer than you need to wait for the style to load, there's little we can do besides moving the map to a hardcoded position that's not the null island while it initializes. I believe that Maps SDK now also respects a https://docs.mapbox.com/mapbox-gl-js/style-spec/root/#center (which it didn't in the past), so that could be the default that most users will fall back to automatically, as long as a style defines it.

**Comment by kmadsen** [comment]
``` kotlin
            // if this is the first location update the activity has received,
            // it's best to immediately move the camera to the current user location
            if (!firstLocationUpdateReceived) {
                firstLocationUpdateReceived = true
                navigationCamera.requestNavigationCameraToOverview(
                    stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
                        .maxDuration(0) // instant transition
                        .build()
                )
            }
```

^ Why not handle this in the `NavigationCamera`? Everyone that implements the navigation camera will need to do this right? I'm confused why the preference is to always start in the ocean 🤔

**Comment by kmadsen** [comment]
As per the recommendation, this is what it is downstream

``` kotlin
    private val locationObserver = object : LocationObserver {
        override fun onRawLocationChanged(rawLocation: Location) {
            // not handled
        }

        override fun onEnhancedLocationChanged(
            enhancedLocation: Location,
            keyPoints: List<Location>
        ) {
            // TODO if you come here looking to fix the initial transition
            //  Look here first https://github.com/mapbox/mapbox-navigation-android/issues/4679
            viewportDataSource.onLocationChanged(enhancedLocation)
            viewportDataSource.evaluate()
            if (!jumpedToFirstLocation) {
                jumpedToFirstLocation = true
                when (cameraMode) {
                    CameraMode.IDLE -> navigationCamera.requestNavigationCameraToIdle()
                    CameraMode.FOLLOWING -> navigationCamera.requestNavigationCameraToFollowing(
                        stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
                            .maxDuration(0) // instant transition
                            .build()
                    )
                    CameraMode.OVERVIEW -> navigationCamera.requestNavigationCameraToOverview(
                        stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
                            .maxDuration(0) // instant transition
                            .build()
                    )
                }
            } else {
                when (cameraMode) {
                    CameraMode.IDLE -> navigationCamera.requestNavigationCameraToIdle()
                    CameraMode.FOLLOWING -> navigationCamera.requestNavigationCameraToFollowing()
                    CameraMode.OVERVIEW -> navigationCamera.requestNavigationCameraToOverview()
                }
            }
        }
    }
```

**Comment by LukasPaczos** [comment]
> I'm confused why the preference is to always start in the ocean

The map has to initialize _somewhere_ if it has no data.

> Why not handle this in the NavigationCamera

I'm worried that it might be surprising to the developer that sometimes `requestNavigationCameraTo_` would ignore the options that were passed to the method and decide to make an instant jump and sometimes it would respect the options. What if a developer manually transitioned the camera already? `NavigationCamera` has no way of knowing that and would jump anyway even though it could transition smoothly.

In the past, we've implemented a system that looks at "pixel distance" (depending on zoom level) between the current camera position and new camera position, and if that distance exceeded a threshold, an immediate jump was performed. We could rebuild an iteration of that approach here as well. This type of configuration could be exposed via `NavigationCameraOptions` to make it accessible.

**Comment by LukasPaczos** [comment]
> As per the recommendation, this is what it is downstream

Do you know which of `onEnhancedLocationChanged` and `onStyleLoaded` fires first? The experience could be different depending on this timing, and this timing could change with every run. You could either wait for the location to be available before starting to load the style, or hardcode an initial camera position. Both have significant drawbacks.

**Comment by kmadsen** [comment]
The NavigationCamera knows when it is initialized or not. It is initializing itself with a valid (0,0) location and it may appear as if it has no way. But instead we're putting it on all downstream developers to figure out. So that is what I'm calling confusing and making it more difficult to use

Can we chew on this and think of some possibilities for improving the `NavigationCamera` interface? But otherwise it seems the recommendation is here https://github.com/mapbox/mapbox-navigation-android/issues/4679#issuecomment-893609306

> Do you know which of onEnhancedLocationChanged and onStyleLoaded fires first? The experience could be different depending on this timing, and this timing could change with every run. You could either wait for the location to be available before starting to load the style, or hardcode an initial camera position. Both have significant drawbacks.

It's not this issue. This situation happens when the style is guaranteed to be loaded, and then we're attaching the map.

I know this is a common problem in maps. For example, at Lyft we would display a foggy screen over the map to hide the initialization. But the nice thing about working for the mapping company, is you get to control exactly what the map is doing.

**Comment by LukasPaczos** [comment]
> It is initializing itself with a valid (0,0)

Could you link to the code where this is happening? I think the `(0,0)` you see is coming from the Map, not from the `NavigationCamera`. The first transition that the `NavigationCamera` executes is already with the coordinates that you provide.

**Comment by kmadsen** [comment]
> I think the (0,0) you see is coming from the Map

Yeah sorry for the late edit on my comment above. I also think the 0,0 is coming from the map. Should we ask the maps team to expose a way for the `NavigationCamera` to know if the map needs to be initialized?

**Comment by kmadsen** [comment]
Also the code we have as an example with this OP issue is

`MapboxCameraAnimationsActivity`
`ReplayHistoryActivity` on this one https://github.com/mapbox/mapbox-navigation-android/pull/4687

**Comment by LukasPaczos** [comment]
> Should we ask the maps team to expose a way for the NavigationCamera to know if the map needs to be initialized?

Even if we had that information, what would we initialize with? The only option would be hardcoding a value, and I don't really see a reasonable value to hardcode. Do you have an idea?

**Comment by kmadsen** [comment]
> Even if we had that information, what would we initialize with? The only option would be hardcoding a value, and I don't really see a reasonable value to hardcode. Do you have an idea?

I think we should take our recommended downstream solution and put it inside. So we don't have to guess where the initial location is. Instead, when the first location comes in - jump to that location.

Some different initial options (before the first location) would be zoom, tilt, padding

**Comment by LukasPaczos** [comment]
> Also the code we have as an example with this OP issue is

The example repository is coming and should have those smoothed out :rocket:

**Comment by LukasPaczos** [comment]
How about we add an initialization block to the `MapboxViewportDataSource` so that it doesn't initialize at the null island:
```kotlin
class MapboxNavigationViewportDataSource(
    private val mapboxMap: MapboxMap,
    initialState: CameraOptions
)
```

and introduce an option to jump on the first location that's enabled by default:
```kotlin
class NavigationCameraOptions(
    val instantFirstTransition: Boolean = true
)
```

What do you think?

**Comment by kmadsen** [comment]
> class MapboxNavigationViewportDataSource(
    private val mapboxMap: MapboxMap,
    initialState: CameraOptions
)

It does not seem we will know the initial coordinate at construction time.

was looking into `MapboxNavigationCameraTransition` but couldn't get it to work with the replay setup https://github.com/mapbox/mapbox-navigation-android/pull/4687 🤔

**Comment by zugaldia** [comment]
> There is no reasonable way to determine the initial location that can be fed to the NavigationCamera

Can the developer manually set a default one?

**Comment by LukasPaczos** [comment]
Yes, our current recommendation is to call
```kotlin
viewportDataSource.onLocationChanged(mapMatcherResult.enhancedLocation)
viewportDataSource.evaluate()
```
as soon as developers have access to a valid location update (or they can pass in a hardcoded default on startup).

**Comment by Guardiola31337** [comment]
Cc @n-frolov for visibility
